crypto/deterministic_nonce.c

```diff
@@ -0,0 +1,195 @@
+ * Returns: 1 if successful, or  0 otherwise.
+ */
+static int bits2octets(unsigned char *out, const BIGNUM *q, int qlen_bits,
+                       int rlen, const unsigned char *in, size_t inlen)
+{
+   int ret = 0;
+   BIGNUM *z = BN_new();
+
+   if (z == NULL
+           || !bits2int(z, qlen_bits, in, inlen))
+       goto err;
+
+   /* z2 = z1 mod q (Do a simple subtract, since z1 < 2^qlen_bits) */
+   if (BN_cmp(z, q) >= 0
+           && !BN_usub(z, z, q))
+       goto err;
+
+   ret = int2octets(out, z, rlen);
+err:
+   BN_free(z);
+   return ret;
+}
+
+/*
+ * Setup a KDF HMAC_DRBG object using fixed entropy and nonce data.
+ *
+ * Params:
+ *     digestname The digest name for the HMAC
+ *     entropy, entropylen A fixed input entropy buffer
+ *     nonce, noncelen A fixed input nonce buffer
+ *     libctx, propq Are used for fetching algorithms
+ *
+ * Returns: The created KDF HMAC_DRBG object if successful, or NULL otherwise.
+ */
+static EVP_KDF_CTX *kdf_setup(const char *digestname,
+                              const unsigned char *entropy, size_t entropylen,
+                              const unsigned char *nonce, size_t noncelen,
+                              OSSL_LIB_CTX *libctx, const char *propq)
+{
+    EVP_KDF_CTX *ctx = NULL;
+    EVP_KDF *kdf = NULL;
+    OSSL_PARAM params[5], *p;
+
+    kdf = EVP_KDF_fetch(libctx, "HMAC-DRBG-KDF", propq);
+    ctx = EVP_KDF_CTX_new(kdf);
+    EVP_KDF_free(kdf);
+    if (ctx == NULL)
+        goto err;
+
+    p = params;
+    *p++ = OSSL_PARAM_construct_utf8_string(OSSL_KDF_PARAM_DIGEST,
+                                            (char *)digestname, 0);
+    if (propq != NULL)
+        *p++ = OSSL_PARAM_construct_utf8_string(OSSL_KDF_PARAM_PROPERTIES,
+                                                (char *)propq, 0);
+    *p++ = OSSL_PARAM_construct_octet_string(OSSL_KDF_PARAM_HMACDRBG_ENTROPY,
+                                             (void *)entropy, entropylen);
+    *p++ = OSSL_PARAM_construct_octet_string(OSSL_KDF_PARAM_HMACDRBG_NONCE,
+                                             (void *)nonce, noncelen);
+    *p = OSSL_PARAM_construct_end();
+
+    if (EVP_KDF_CTX_set_params(ctx, params) <= 0)
+        goto err;
+
+    return ctx;
+err:
+    EVP_KDF_CTX_free(ctx);
+    return NULL;
+}
+
+/*
+ * Generate a Deterministic nonce 'k' for DSA/ECDSA as defined in
+ * RFC 6979 Section 3.3.  "Alternate Description of the Generation of k"
+ *
+ * Params:
+ *     out Returns the generated deterministic nonce 'k'
+ *     q A large prime number used for modulus operations for DSA and ECDSA.
+ *     priv The private key in the range [1, q-1]
+ *     hm, hmlen The digested message buffer in bytes
+ *     digestname The digest name used for signing. It is used as the HMAC digest.
+ *     libctx, propq Used for fetching algorithms
+ *
+ * Returns: 1 if successful, or  0 otherwise.
+ */
+int ossl_gen_deterministic_nonce_rfc6979(BIGNUM *out, const BIGNUM *q,
```

Comment:
I would have expected this to be conditioned out in a FIPS module build.

Reply:
It is done via the build.info as are quite a few files in /crypto